### PR TITLE
Refactor key interaction tests - Pass 1

### PR DESCRIPTION
**Pull request by aicioara**
Part of #2628

I will be doing 2 passes because I will get more context (and better sense of best practices) by refactoring other files. However, please flag anything that seems unusual.

As part of this: 
- **Test semantics: hierarchy, naming, testing, but no coverage check**
- Applied best practices from the wiki page
- Factored out common code into `beforeEach()` clauses
- No nested `beforeEach()`es
- Cleaned the console
- Remade hierarchy as shown bellow
  screen shot 2015-09-11 at 12 35 35 pm

### CODE REVIEW

File: test/interactions/keyInteractionTests.ts [hunk 1]
```diff
@@ -1,34 +1,35 @@
...
+      let svg: d3.Selection<void>;
+      let component: Plottable.Component;
+      let keyInteraction: Plottable.Interactions.Key;
+      let aCode: number;
+      let bCode: number;
+
+      let aCallbackCalled: boolean;
+      let bCallbackCalled: boolean;
+      let aCallback: Plottable.KeyCallback;
+      let bCallback: Plottable.KeyCallback;
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Why we need the `b` version of these version

File: test/interactions/keyInteractionTests.ts [hunk 2]
```diff
@@ -75,7 +77,7 @@ describe("Interactions", () => {
         svg.remove();
       });
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: The "can remove keyPress callbacks" should assert that the function returns the calling object.

File: test/interactions/keyInteractionTests.ts [hunk 3]
```diff
@@ -75,7 +77,7 @@ describe("Interactions", () => {
...
 
-      it("multiple keyPress callbacks are possible", () => {
+      it("can attach multiple multiple keyPress callbacks", () => {
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: double "multiple"?  Remove one of them

Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: If we have this, what's the use of `bCallback`?

File: test/interactions/keyInteractionTests.ts (hunk 1 above)
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Can we have better names for these variables instead of a/b?

File: test/interactions/keyInteractionTests.ts [hunk 4]
```diff
@@ -149,14 +151,40 @@ describe("Interactions", () => {
...
+      let svg: d3.Selection<void>;
+      let component: Plottable.Component;
+      let keyInteraction: Plottable.Interactions.Key;
+      let aCode: number;
+      let bCode: number;
+
+      let aCallbackCalled: boolean;
+      let bCallbackCalled: boolean;
+      let aCallback: Plottable.KeyCallback;
+      let bCallback: Plottable.KeyCallback;
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Same as above concerns.

File: test/interactions/keyInteractionTests.ts [hunk 5]
```diff
@@ -230,7 +258,7 @@ describe("Interactions", () => {
         svg.remove();
       });
 
-      it("removing keyRelease callbacks is possible", () => {
+      it("can remove removing keyRelease callbacks", () => {
         keyInteraction.onKeyRelease(aCode, aCallback);
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: We should assert that calling this function returns the calling object.

**Comment by bluong** [comment]
Main concern being the usage of the `aCallback` and `bCallback` variables.  I think it would make more sense to have a single callback at the describe level and then declare the other variables as needed in the unit tests.

File: test/interactions/keyInteractionTests.ts (hunk 2 above)
Comment (aicioara) [inline test/interactions/keyInteractionTests.ts]: I do that already in the first `it`

File: test/interactions/keyInteractionTests.ts (hunk 3 above)
Comment (aicioara) [inline test/interactions/keyInteractionTests.ts]: `bCallback` is for key "b"

File: test/interactions/keyInteractionTests.ts (hunk 5 above)
Comment (aicioara) [inline test/interactions/keyInteractionTests.ts]: That is being asserted in the previous "it". Does it make sense to only assert it here?

**Comment by aicioara** [comment]
Fixed the `aCallback` `bCallback` issue

File: test/interactions/keyInteractionTests.ts (hunk 2 above)
Comment (aicioara) [inline test/interactions/keyInteractionTests.ts]: I actually moved it here now as it makes more sense.

File: test/interactions/keyInteractionTests.ts [hunk 6]
```diff
@@ -1,302 +1,327 @@
...
+        component = new Plottable.Component();
+        keyInteraction = new Plottable.Interactions.Key();
+        aKeyCode = 65;
+        aKeyCodeCallbackCalled = false;
+        aKeyCodeCallback = () => aKeyCodeCallbackCalled = true;
+        svg = TestMethods.generateSVG();
+        component.renderTo(svg);
+      });
+
       it("only fires callback for \"a\" when \"a\" key is pressed", () => {
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: This does not sounds like the type of test that should be at the front of this test suite.

Instead, can we have something like
it("fires callback for the a key code when that key is pressed)
as the first test and then have
it("does not fire callbacks for key codes other than the pressed key")
as an edge case second test

Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: As a result we can move the setter assertion up to this new first test here.

File: test/interactions/keyInteractionTests.ts [hunk 7]
```diff
@@ -1,302 +1,327 @@
...
-        keyInteraction.onKeyPress(aCode, aCallback);
-        TestMethods.triggerFakeKeyboardEvent("keydown", component.background(), aCode);
-        assert.isTrue(aCallbackCalled, "callback for \"a\" was properly connected back to the interaction");
+        keyInteraction.onKeyPress(aKeyCode, aKeyCodeCallback);
+        TestMethods.triggerFakeKeyboardEvent("keydown", component.background(), aKeyCode);
+        assert.isTrue(aKeyCodeCallbackCalled, "callback for \"a\" was properly connected back to the interaction");
 
-        keyInteraction.offKeyPress(aCode, aCallback);
+        assert.strictEqual(keyInteraction.offKeyPress(aKeyCode, aKeyCodeCallback), keyInteraction,
+          "unsetting the keyPress callback returns the interaction");
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: It seems that this test is essentially divided into two parts...  What is the second part trying to test?

Also it seems like the last line here is testing that the setter is chainable; shouldn't this assert be made higher up?

File: test/interactions/keyInteractionTests.ts [hunk 8]
```diff
@@ -1,302 +1,327 @@
...
         svg.remove();
       });
 
-      it("multiple keyPress callbacks are possible", () => {
-        let aCallback1Called = false;
-        let aCallback1 = () => aCallback1Called = true;
-        let aCallback2Called = false;
-        let aCallback2 = () => aCallback2Called = true;
+      it("can attach multiple keyPress callbacks", () => {
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: let's be more descriptive and append
" to the same key code" or " to the same key"

File: test/interactions/keyInteractionTests.ts [hunk 9]
```diff
@@ -1,302 +1,327 @@
...
+        assert.isTrue(aKeyCodeCallback2Called, "callback 2 for \"a\" was called when \"a\" key was pressed");
 
-        keyInteraction.offKeyPress(aCode, aCallback1);
-        keyInteraction.offKeyPress(aCode, aCallback2);
+        keyInteraction.offKeyPress(aKeyCode, aKeyCodeCallback1);
+        keyInteraction.offKeyPress(aKeyCode, aKeyCodeCallback2);
         svg.remove();
       });
 
       it("can remove only one of the registered keypress callbacks", () => {
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Maybe rename to
it("removes only the specified callback")?

File: test/interactions/keyInteractionTests.ts [hunk 10]
```diff
@@ -1,302 +1,327 @@
...
+        aKeyCodeCallbackCalled = false;
+        TestMethods.triggerFakeKeyboardEvent("keydown", component.background(), aKeyCode, { repeat : true });
+        assert.isFalse(aKeyCodeCallbackCalled, "callback for \"a\" was not called when keydown was fired the second time");
 
-        keyInteraction.offKeyPress(aCode, aCallback);
+        keyInteraction.offKeyPress(aKeyCode, aKeyCodeCallback);
         svg.remove();
       });
 
       it("does not fire the callback if the key is pressed and held down outside, then the mouse moved inside", () => {
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: This sentence is quite a bit long and a bit nondescriptive ("outside... what?")

File: test/interactions/keyInteractionTests.ts [hunk 11]
```diff
@@ -1,302 +1,327 @@
...
+        aKeyCode = 65;
+        component = new Plottable.Component();
+        keyInteraction = new Plottable.Interactions.Key();
+        aKeyCodeCallbackCalled = false;
+        aKeyCodeCallback = () => aKeyCodeCallbackCalled = true;
+        svg = TestMethods.generateSVG();
+        component.renderTo(svg);
+      });
+
       it("doesn't fire callback if key was released without being pressed", () => {
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Same as above how we should just have a base unit test about setting the `onKeyRelease` callback

**Comment by bluong** [comment]
Few test naming issues and we should add in the base unit tests.

File: test/interactions/keyInteractionTests.ts (hunk 7 above)
Comment (aicioara) [inline test/interactions/keyInteractionTests.ts]: Moved it higher up. It was a mistake to make this assert here. The test was basically checking that the callback can be removed and reattahed

Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Can we put the reattach in a different test?

File: test/interactions/keyInteractionTests.ts (hunk 6 above)
Comment (aicioara) [inline test/interactions/keyInteractionTests.ts]: Fixed

File: test/interactions/keyInteractionTests.ts (hunk 7 above)
Comment (aicioara) [inline test/interactions/keyInteractionTests.ts]: I think that is a bit overkill. The reattach test will basically copy + paste this entire testsuite. I totally agree with the modular approach, but the benefits of having this vs the extra test + time to run the testsuite is not worth it, hence I suggest we leave it like this

File: test/interactions/keyInteractionTests.ts (hunk 11 above)
Comment (aicioara) [inline test/interactions/keyInteractionTests.ts]: FIxed

File: test/interactions/keyInteractionTests.ts [hunk 12]
```diff
@@ -1,302 +1,340 @@
...
-      bCallbackCalled = false;
-    });
-
+  describe("Key Interaction", () => {
     describe("onKeyPress", () => {
+      let svg: d3.Selection<void>;
+      let component: Plottable.Component;
+      let keyInteraction: Plottable.Interactions.Key;
+
+      let aKeyCode: number;
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Technically this should be `const`ed, or we can even capitalize case it.  What do you think?

File: test/interactions/keyInteractionTests.ts [hunk 13]
```diff
@@ -1,302 +1,340 @@
...
+      it("fires callback for key press", () => {
+        assert.strictEqual(keyInteraction.onKeyPress(aKeyCode, aKeyCodeCallback), keyInteraction,
+          "setting the keyPress callback returns the interaction");
+        keyInteraction.attachTo(component);
+
+        TestMethods.triggerFakeMouseEvent("mouseover", component.background(), 100, 100);
+        TestMethods.triggerFakeKeyboardEvent("keydown", component.background(), aKeyCode);
+        assert.isTrue(aKeyCodeCallbackCalled, "callback for \"a\" was called when \"a\" key was pressed");
+
+        keyInteraction.offKeyPress(aKeyCode, aKeyCodeCallback);
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Why this line?

File: test/interactions/keyInteractionTests.ts (hunk 2 above)
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Do you think it would be a simpler yet sufficient test if we instead just test if "bCallback" is not called if "a" is pressed?  What I mean by this is not to touch the setup for "a" at all.

File: test/interactions/keyInteractionTests.ts [hunk 14]
```diff
@@ -1,302 +1,340 @@
...
         keyInteraction.attachTo(component);
 
         TestMethods.triggerFakeMouseEvent("mouseout", component.background(), -100, -100);
-        TestMethods.triggerFakeKeyboardEvent("keydown", component.background(), aCode);
-        assert.isFalse(aCallbackCalled, "callback for \"a\" was not called when not moused over the Component");
+        TestMethods.triggerFakeKeyboardEvent("keydown", component.background(), aKeyCode);
+        assert.isFalse(aKeyCodeCallbackCalled, "callback for \"a\" was not called when not moused over the Component");
 
-        keyInteraction.offKeyPress(aCode, aCallback);
+        keyInteraction.offKeyPress(aKeyCode, aKeyCodeCallback);
```
Comment (bluong) [inline test/interactions/keyInteractionTests.ts]: Why this line?

**Comment by bluong** [comment]
The above new comments pervade through the file in multiple locations so I jsut commented on the first occurrence.

File: test/interactions/keyInteractionTests.ts (hunk 12 above)
Comment (aicioara) [inline test/interactions/keyInteractionTests.ts]: Fixed

**Comment by aicioara** [comment]
Since can't continue on this branch anymore, moving to the fork.
